### ISSUE TITLE
Howto support multiple projects in llama-index

### ISSUE THREAD

**Issue by stdweird** [Arize-ai/phoenix#3375]
I am trying to use phoenix with llama-index and have 2 functions; and i want each function to have its spans send to a dedicated project. So far, i don't succeed in it. E.g. setting the project environment variable as part of each function call is not enough to make this work.

I am looking for more guidance where this could be set, or where i can "manipulate" this whole process. I'm ok with pointers to code, but i would probably have to search llamaindex, openinference and phoenix code; so some help is welcome

**Comment by mikeldking** [comment]
Hey @stdweird - Thanks for the question and thanks for trying out Phoenix.

Just wanted to double check if you had taken a look at some of our docs on the topic (https://docs.arize.com/phoenix/tracing/how-to-tracing/customize-traces)

That being said I totally understand your request. In general Phoenix projects are designed to be specified at an application level so our standard recommendation is to specify it at the resource level:

```python
from openinference.semconv.resource import ResourceAttributes
from openinference.instrumentation.llama_index import LlamaIndexInstrumentor
from opentelemetry import trace as trace_api
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk import trace as trace_sdk
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace.export import SimpleSpanProcessor

resource = Resource(attributes={
    ResourceAttributes.PROJECT_NAME: '<your-project-name>'
})
tracer_provider = trace_sdk.TracerProvider(resource=resource)
span_exporter = OTLPSpanExporter(endpoint="http://phoenix:6006/v1/traces")
span_processor = SimpleSpanProcessor(span_exporter=span_exporter)
tracer_provider.add_span_processor(span_processor=span_processor)
trace_api.set_tracer_provider(tracer_provider=tracer_provider)
# Add any auto-instrumentation you want 
LlamaIndexInstrumentor().instrument()
```

That being said I see you would like to split two sub-parts of your existing application into two different projects.

Before I point out a possible solution for this have you considered segmenting these spans using metadata rather than projects? We have a section on how to add metadata to spans that are auto-instrumented (https://docs.arize.com/phoenix/tracing/how-to-tracing/customize-spans)

I think this would be the preferred way for us as it would still group all your spans for a given application under 1 project but would let you filter down to spans by metadata of your choosing.

**Comment by stdweird** [comment]
@mikeldking i'll have a look at the metadata, but the "2 functions" is an oversimplification of the actual code: it's a multipage streamlit app, where each page has nothing to do with the others. so yes, i prefer multiple projects (we will use the metadata to tag the users etc)

**Comment by mikeldking** [comment]
> @mikeldking i'll have a look at the metadata, but the "2 functions" is an oversimplification of the actual code: it's a multipage streamlit app, where each page has nothing to do with the others. so yes, i prefer multiple projects (we will use the metadata to tag the users etc)

@stdweird gotcha. Well, I do have a solution for you - it's a bit non-standard but we do have a way to use a context manager to switch projects within a single app. We typically don't recommend this outside of notebooks but it does work (it overrides the resource attributes of the exporters). See source code here (https://github.com/Arize-ai/phoenix/blob/68d24f9a395887a66a0e3a8395e884c5b17710b8/src/phoenix/trace/projects.py#L29)

I think this is the only work around I have for your right now. LMK if this works for you. Feel free to just lift this bit of code so you don't have to depend on phoenix to just switch projects.

**Comment by mikeldking** [comment]
> @mikeldking thanks, i'll give the code a try right away. does setting the project via the env variable do anything at all in this context (i guess it doesn't hurt either)

@stdweird the env variable is a way to have everything in the phoenix namespace use that project by default. So things like downloading traces, sending evals etc. all defaults to that project. But it doesn't impact the instrumentors if they are initialized the "OTel" way as specified above.

TLDR - If you use the code above, you won't have to set the env var.

**Comment by stdweird** [comment]
@mikeldking wow, that just works!!!

anyway, feel free to move the code to the opentelemetry repo, so i can import it directly ;)

**Comment by axiomofjoy** [comment]
@stdweird Is this issue resolved? Good to close?

**Comment by stdweird** [comment]
yes. sorry about not clsing sooner

**Comment by axiomofjoy** [comment]
No worries, thanks!
